subsys/disk/disk_access_sdhc.c

```diff
@@ -0,0 +1,1001 @@
+	err = sdhc_cmd_r1(data, SDHC_READ_MULTIPLE_BLOCK, sector);
+	if (err != 0) {
+		goto error;
+	}
+
+	/* Read the sectors */
+	for (; count != 0; count--) {
+		err = sdhc_rx_block(data, buf, SDHC_SECTOR_SIZE);
+		if (err != 0) {
+			goto error;
+		}
+		buf += SDHC_SECTOR_SIZE;
+	}
+
+	/* Ignore the error as STOP_TRANSMISSION always returns 0x7F */
+	sdhc_cmd_r1(data, SDHC_STOP_TRANSMISSION, 0);
+
+	/* Wait until the card becomes ready */
+	err = sdhc_skip_until_ready(data);
+
+error:
+	sdhc_set_cs(data, 1);
+
+	return err;
+}
+
+static int sdhc_write(struct sdhc_data *data, const u8_t *buf, u32_t sector,
+		      u32_t count)
+{
+	int err;
+
+	err = sdhc_map_disk_status(data->status);
+	if (err != 0) {
+		return err;
+	}
+
+	sdhc_set_cs(data, 0);
+
+	/* Write the blocks one-by-one */
+	for (; count != 0; count--) {
+		err = sdhc_cmd_r1(data, SDHC_WRITE_BLOCK, sector);
+		if (err < 0) {
+			goto error;
+		}
+
+		err = sdhc_tx_block(data, (u8_t *)buf, SDHC_SECTOR_SIZE);
+		if (err != 0) {
+			goto error;
+		}
+
+		/* Wait for the card to finish programming */
+		err = sdhc_skip_until_ready(data);
+		if (err != 0) {
+			goto error;
+		}
+
+		err = sdhc_cmd_r2(data, SDHC_SEND_STATUS, 0);
+		if (err != 0) {
+			goto error;
+		}
+
+		buf += SDHC_SECTOR_SIZE;
+		sector++;
+	}
+
+	err = 0;
+error:
+	sdhc_set_cs(data, 1);
+
+	return err;
+}
+
+static int sdhc_init(struct device *dev)
+{
+	struct sdhc_data *data = dev->driver_data;
+
+	data->cfg.dev = device_get_binding(CONFIG_SPI_0_NAME);
+	__ASSERT_NO_MSG(data->cfg.dev != NULL);
+
+	data->cfg.frequency = SDHC_INITIAL_SPEED;
+	data->cfg.operation = SPI_WORD_SET(8) | SPI_HOLD_ON_CS;
+
+	data->cs = device_get_binding(CONFIG_SPI_0_CS_GPIO_PORT);
+	__ASSERT_NO_MSG(data->cs != NULL);
+
+	data->pin = CONFIG_SPI_0_CS_GPIO_PIN;
+
+	return gpio_pin_configure(data->cs, data->pin, GPIO_DIR_OUT);
+}
+
+static struct device *sdhc_get_device(void) { return DEVICE_GET(sdhc_0); }
+
+int disk_access_status(void)
+{
+	struct device *dev = sdhc_get_device();
+	struct sdhc_data *data = dev->driver_data;
+
+	return data->status;
+}
+
+int disk_access_read(u8_t *buf, u32_t sector, u32_t count)
+{
+	struct device *dev = sdhc_get_device();
+	struct sdhc_data *data = dev->driver_data;
+	int err;
+
+	SYS_LOG_DBG("sector=%u count=%u", sector, count);
+
+	err = sdhc_read(data, buf, sector, count);
+
+	if (err != 0 && sdhc_is_retryable(err)) {
+		sdhc_recover(data);
+		err = sdhc_read(data, buf, sector, count);
+	}
+
+	return err;
+}
+
+int disk_access_write(const u8_t *buf, u32_t sector, u32_t count)
+{
+	struct device *dev = sdhc_get_device();
+	struct sdhc_data *data = dev->driver_data;
+	int err;
+
+	SYS_LOG_DBG("sector=%u count=%u", sector, count);
+
+	err = sdhc_write(data, buf, sector, count);
+
+	if (err != 0 && sdhc_is_retryable(err)) {
+		sdhc_recover(data);
+		err = sdhc_write(data, buf, sector, count);
+	}
+
+	return err;
+}
+
+int disk_access_ioctl(u8_t cmd, void *buf)
+{
+	struct device *dev = sdhc_get_device();
```

Comment:
this disk access api seems rather limited: only 1 disk is possible to get...

Reply:
Agreed, the Zephyr API is limited.  Longer term, I think Zephyr should switch to a block device driver interface so you can support, say, config in NFFS on internal flash and logging on FAT on a SD card.

Anything you'd like me to do here?

Reply:
we have a issue dealing with this and someone looking at it.